pkg/plugins/runtime/k8s/webhooks/injector/injector.go

```diff
@@ -129,21 +129,57 @@ func (i *KumaInjector) needInject(pod *kube_core.Pod, ns *kube_core.Namespace) (
 		log.V(1).Info("pod fulfills exception requirements")
 		return false, nil
 	}
-	enabled, exist, err := metadata.Annotations(pod.Annotations).GetEnabled(metadata.KumaSidecarInjectionAnnotation)
+
+	for _, container := range pod.Spec.Containers {
+		if container.Name == util.KumaSidecarContainerName {
+			log.V(1).Info("pod already has Kuma sidecar")
+			return false, nil
+		}
+	}
+
+	enabled, exist, err := metadata.Annotations(pod.Labels).GetEnabled(metadata.KumaSidecarInjectionAnnotation)
 	if err != nil {
 		return false, err
 	}
 	if exist {
+		if !enabled {
+			log.V(1).Info("pod has kuma.io/sidecar-injection: disabled label")
+		}
+		return enabled, nil
+	}
+
+	// support annotations for backwards compatibility
+	annotationWarningMsg := "WARNING: you are using kuma.io/sidecar-injection as annotation. Please migrate it to label to have strong guarantee that application can only start with sidecar"
```

Comment:
Why doesn't kuma migrate the labels?

Reply:
Technically we could, but you most likely store this in some git repo that applies YAMLs on Kubernetes.

I'd like to encourage users to update this in their single source of truth.